### ISSUE TITLE
csm_allocation_query_details missing gpu_usage

### ISSUE THREAD

**Issue by morrone** [IBM/CAST#533]
In the output of the "csm_allocation_query_details -a <allocation id>" command, I am finding that some nodes are missing the "gpu_usage" field.

The very same allocation will list gpu_usage for some of the nodes, but not for other of the nodes.  Further, the nodes without gpu_usage fields do have gpu_energy fields, and the gpu_energy of the nodes without gpu_usage are similar in value to the gpu_energy on nodes with gpu_usage.

This makes me wonder if this is a bug rather than, say, the field being omitted because the value was 0 for those nodes.  It seems unlikely that we could have zero gpu_usage when the gpu_energy is on par with other nodes in the same allocation.

**Comment by besawn** [comment]
There are a few different elements to this issue, so I will address them individually.

**1.) The first aspect is that "csm_allocation_query_details" does not print certain fields (including "gpu_usage") if they are less than 1.**

This particular aspect of the issue is a duplicate of issue #473. That issue has already been fixed and the fix will be included with the next CSM release, currently planned to be CSM 1.5.0.

**2.) The second aspect is distinguishing between nodes that are reporting a value of 0 for "gpu_usage" vs nodes that encountered an error while trying to collect "gpu_usage".**

If you would like to debug this aspect further, this can be debugged by examining the actual values for "gpu_usage" stored in the "csm_allocation_node_history" of the CSM DB. For nodes that have encountered an error during collection, "gpu_usage" will be set to -1; nodes that have a value of 0 for "gpu_usage" are reporting the "gpu_usage" to be 0.

If you have superuser access on the node where the CSM DB is running, something like this should work: 
Note: replace allocation_id='1' with the relevant allocation_id value.
```
/bin/psql -U postgres csmdb -c "select allocation_id,state,node_name,gpu_energy,gpu_usage from csm_allocation_node_history where allocation_id='1' AND state='complete'"
 allocation_id |  state   | node_name  | gpu_energy | gpu_usage 
---------------+----------+------------+------------+-----------
             1 | complete | c650f99p18 |      38023 | 197694624
             1 | complete | c650f99p26 |      39323 | 197719940
             1 | complete | c650f99p28 |      39172 | 197679820
             1 | complete | c650f99p36 |      58878 | 296535762
```

This aspect of the issue is also improved with the fix for issue #473, as either -1 or 0 will be reported in the "csm_allocation_query_details" output, starting in CSM 1.5.0.

**3.) The third aspect is related to the question of whether different nodes in an allocation should report values with large variance for "gpu_usage" when they report similar values for "gpu_energy".**

`It seems unlikely that we could have zero gpu_usage when the gpu_energy is on par with other nodes in the same allocation.`

This aspect is related to differences in how "gpu_usage" and "gpu_energy" are collected via the GPU driver. "gpu_energy" can be thought of as a monotonically increasing value that is always accurate whenever it is read. On the other hand, "gpu_usage" is a derived value that is calculated based on an underlying field that is sampled periodically and averaged. Due to the differences in how these fields are constructed, it is quite likely that large variations can be observed in reported "gpu_usage" for similar values of "gpu_energy".

The accuracy of "gpu_usage" is related to characteristics of the job, length of the allocation, sampling frequency, total number of samples, etc.

By default, CSM samples the gpu_utilization every 30 seconds and then calculates the "gpu_usage" using those samples. For allocations that last less than 5 minutes, the "gpu_usage" is calculated on ~9 samples. The default value can be changed, but this involves a trade off between the accuracy of the measured "gpu_usage" and overhead of collecting many samples during the life of an allocation. 

In the most extreme case, for a job that runs for ~30 seconds that spends half of the time idle and half of the time at 100% gpu utilization, each node would be expected to take 1 sample, and the sample would always be either 0% utilization or 100% utilization. This sampling error would then propagate through to the calculation of "gpu_usage".

In summary:

- If aspect 1 is the main concern, I think this issue can be closed as a duplicate of #473.

- If aspect 2 is a concern, we can verify that none of the nodes are reporting a "gpu_usage" of -1, which would indicate an error occurred.

- If aspect 3 is a concern, I think this is most likely expected behavior for allocations that last less than 5 minutes with the default sampling frequency of 30 seconds. It is possible to change this sampling frequency, but it involves a trade off. This specific tuning item has also been heavily discussed on past interlock calls, so if making a change to the tuning seems desirable, this might warrant revisiting the tuning discussion again with all interested parties.

**Comment by morrone** [comment]
Well, the sampling rate does seem to be wildly inaccurate in practice (assuming for the moment that "aspect 2" is not at play).  I think that sampling more frequently might be required.  Keep in mind that we do not need to necessarily store all samples.  We can use averaging at the point of collection to keep the recorded data lower while still providing better insight into gpu usage.

**Comment by morrone** [comment]
FYI, our recollection is that all of the interested parties agreed that we would need to sample the gpu usage at a minimum of once per second.  If discussion is necessary to make that work well, then yes, lets schedule that discussion.

**Comment by morrone** [comment]
Also, thanks for the psql command example.  I have verified that the gpu_usage is 0 rather than -1 in a couple of cases where I hit missing gpu_usage output from csm_allocation_query_details.  So that part seems clear.

"aspect 3" then is our main concern at the moment.

**Comment by besawn** [comment]
For aspect 3, the default sampling frequency of 30 seconds can be modified be adding the following block to the end of the `/etc/ibm/csm/csm_compute.cfg` file on each compute node:
```
        },

        "tuning" :
        {
                "dcgm_update_interval_s" : 1
        }
```
_**Make sure to include the comma between the "data_collection" and "tuning" sections when adding the "tuning" section to the file.**_

Before:
```
        "data_collection" :
        {
            "buckets":
                [
                    {
                        "execution_interval":"00:10:00",
                        "item_list": ["environmental"]
                    }
                ]
        }
    }
}
```

After:
```
        "data_collection" :
        {
            "buckets":
                [
                    {
                        "execution_interval":"00:10:00",
                        "item_list": ["environmental"]
                    }
                ]
        },

        "tuning" :
        {
                "dcgm_update_interval_s" : 1
        }
    }
}
```

"dcgm_update_interval_s" is specified in seconds.
The minimum supported value for "dcgm_update_interval_s" is 1 second.
The default value is 30 seconds.

Using a smaller "dcgm_update_interval_s" results in greater accuracy when measuring gpu_usage at the cost of additional CPU and memory usage by the nv-hostengine process that is collecting the data.

**Comment by besawn** [comment]
I think this issue has been resolved several releases ago, please re-open if it is still a problem.
